File: 백준/Platinum/33807. Minimum Spanning Arborescence/Minimum Spanning Arborescence.py
import sys
input = lambda: sys.stdin.readline().rstrip()
print = lambda *args, sep=" ", end="\n": sys.stdout.write(sep.join(map(str, args)) + end)

def main():
    MOD = 998244353
    n, m, k = map(int, input().split())
    deg = [0] * (n + 1)
    for _ in range(m):
        _, y = map(int, input().split())
        deg[y] += 1

    vcnt = [0] * (m + 1)
    maxd = 0
    for i in range(2, n + 1):
        if deg[i]:
            vcnt[deg[i]] += 1
            if deg[i] > maxd:
                maxd = deg[i]

    if maxd == 0:
        print(0)
        return

    # ── 1. 역팩토리얼 precompute ──────────────────────────────────────
    size = maxd + 3
    fact = [1] * size
    for i in range(1, size):
        fact[i] = fact[i-1] * i % MOD
    inv_fact = [1] * size
    inv_fact[size-1] = pow(fact[size-1], MOD-2, MOD)
    for i in range(size-2, -1, -1):
        inv_fact[i] = inv_fact[i+1] * (i+1) % MOD

    # ── 2. k의 falling factorial & 역수 precompute ───────────────────
    # fall[i] = k*(k-1)*...*(k-i)  (i+1개 항의 곱)
    fall_size = maxd + 2  # 인덱스 0..maxd+1
    fall = [0] * fall_size
    fall[0] = k % MOD
    for i in range(1, fall_size):
        fall[i] = fall[i-1] * ((k - i) % MOD) % MOD

    # inv_fall[i] = 1/fall[i]  (배치 역원: modpow 1회)
    inv_fall = [0] * fall_size
    inv_fall[-1] = pow(fall[-1], MOD-2, MOD)
    for i in range(fall_size-2, -1, -1):
        inv_fall[i] = inv_fall[i+1] * ((k - i - 1) % MOD) % MOD

    # inv_kj[i] = 1/(k-i)
    #   = fall[i-1] * inv_fall[i]  (i>=1),  inv_fall[0]  (i=0)
    inv_kj = [0] * fall_size
    inv_kj[0] = inv_fall[0]
    for i in range(1, fall_size):
        inv_kj[i] = fall[i-1] * inv_fall[i] % MOD

    # ── 3. invk^d precompute ─────────────────────────────────────────
    invk = pow(k % MOD, MOD-2, MOD)
    invk_pow = [1] * (maxd + 1)
    for i in range(1, maxd + 1):
        invk_pow[i] = invk_pow[i-1] * invk % MOD

    # ── Lagrange 보간으로 sum_{i=1}^{k} i^d 계산 ─────────────────────
    def Sd(d):
        pts = d + 1 if k > d + 1 else k
        y = [0] * (pts + 1)
        for i in range(1, pts + 1):
            y[i] = (y[i-1] + pow(i, d, MOD)) % MOD
        if k <= d + 1:
            return y[k]

        # 라그랑주 보간: x=k, 보간점 (0,y[0])...(d+1,y[d+1])
        # numerator_i = fall[d+1] / (k-i)  ← pref/suff 배열 불필요
        P = fall[d + 1]
        ret = 0
        for i in range(d + 2):
            num = P * inv_kj[i] % MOD
            den = inv_fact[i] * inv_fact[d + 1 - i] % MOD
            if (d + 1 - i) % 2:
                den = MOD - den
            ret = (ret + y[i] * num % MOD * den) % MOD
        return ret

    ans = 0
    for d in range(1, maxd + 1):
        if vcnt[d] == 0:
            continue
        evmd = Sd(d) * invk_pow[d] % MOD
        ans = (ans + vcnt[d] * evmd) % MOD

    print(ans)

if __name__ == "__main__":
    main()

File: 백준/Platinum/33807. Minimum Spanning Arborescence/Minimum Spanning Arborescence.cc
#include <bits/stdc++.h>
using namespace std;
using ll = long long;

const ll MOD = 998244353;
int k;
vector<ll> invf, pref, suff, invki;


ll pw(ll n, ll r, ll MOD){
    n %= MOD;
    ll ret = 1;
    while(r){
        if(r&1) ret = ret*n%MOD;
        n = n*n%MOD;
        r >>= 1;
    }
    return ret;
}

ll Sd(int d){
    vector<ll> y(d+2, 0);
    for(int i = 1; i < d+2; i++) y[i] = (y[i-1]+pw(i, d, MOD))%MOD;
    if(k <= d+1) return y[k];
    ll ret = 0;
    for(int i = 0; i < d+2; i++){
        ll num = (pref[d+1]*invki[i])%MOD;
        ll den = (invf[i]*invf[d+1-i])%MOD;
        if((d+1-i)%2) den = MOD-den;
        ret = (ret+y[i]*num%MOD*den)%MOD;
    }
    return ret;
}

int main() {
    cin.tie(0); cout.tie(0); ios::sync_with_stdio(false);
    int n, m; cin >> n >> m >> k;
    vector<int> deg(n+1, 0);
    for(int i = 0; i < m; i++){
        int x, y; cin >> x >> y;
        deg[y]++;
    }
    vector<int> vcnt(m+1, 0);
    int maxd = 0;
    for(int i = 2; i <= n; i++){
        if(deg[i]) vcnt[deg[i]]++;
        if(deg[i] > maxd) maxd = deg[i];
    }
    int limit = min(k-1, maxd+1);
    pref.assign(limit+1, 0), suff.assign(limit+1, 0);
    pref[0] = k;
    for(int i = 1; i <= limit; i++) pref[i] = (pref[i-1]*(k-i))%MOD;
    suff[limit] = pw(pref[limit], MOD-2, MOD);
    for(int i = limit-1; i >= 0; i--) suff[i] = (suff[i+1]*(k-i-1))%MOD;
    invki.assign(limit+1, 0);
    invki[0] = suff[0];
    for(int i = 1; i <= limit; i++) invki[i] = (pref[i-1]*suff[i])%MOD;
    ll fact = 1; invf.assign(maxd+3, 1); 
    for(int i = 1; i < maxd+3; i++) fact = (fact*i)%MOD;
    invf[maxd+2] = pw(fact, MOD-2, MOD);
    for(int i = maxd+1; i > 0; i--) invf[i] = (invf[i+1]*(i+1))%MOD;
    ll ans = 0;
    ll invk = pw(k, MOD-2, MOD);
    vector<ll> invkp(maxd+1, 1);
    for(int i = 1; i <= maxd; i++) invkp[i] = (invkp[i-1]*invk)%MOD;
    for(int d = 1; d <= maxd; d++){
        if(vcnt[d] == 0) continue;
        ll evmd = (Sd(d)*invkp[d])%MOD;
        ans = (ans+vcnt[d]*evmd)%MOD;
    }
    cout << ans;
    return 0;
}